stdlib/public/runtime/ImageInspectionELF.cpp

```diff
@@ -18,144 +18,40 @@
-}
-
-static int iteratePHDRCallback(struct dl_phdr_info *info,
-                               size_t size, void *data) {
-  InspectArgs *inspectArgs = reinterpret_cast<InspectArgs *>(data);
-  const char *fname = info->dlpi_name;
-
-  // While dl_iterate_phdr() is in progress it holds a lock to prevent other
-  // images being loaded. The initialize flag is set here inside the callback so
-  // that addNewDSOImage() sees a consistent state. If it was set outside the
-  // dl_iterate_phdr() call then it could result in images being missed or
-  // added twice.
-  inspectArgs->didInitializeLookup = true;
+extern "C" {
+// Create empty sections to ensure that the start/stop symbols are synthesized
+// by the linker.  Otherwise, we may end up with undefined symbol references as
+// the linker table section was never constructed.
 
-  if (fname == nullptr || fname[0] == '\0') {
-    // The filename may be null for both the dynamic loader and main executable.
-    // So ignore null image name here and explicitly add the main executable
-    // in initialize*Lookup() to avoid adding the data twice.
-    return 0;
-  }
+#define DECLARE_SWIFT_SECTION(name)                                            \
+  __asm__("\t.section " #name ",\"a\"\n");                                     \
+  __attribute__((__visibility__("hidden"))) extern const char __start_##name;  \
+  __attribute__((__visibility__("hidden"))) extern const char __stop_##name;
 
-  SectionInfo block = getSectionInfo(fname, inspectArgs->symbolName);
-  if (block.size > 0) {
-    inspectArgs->addBlock(block.data, block.size);
-  }
-  return 0;
-}
+DECLARE_SWIFT_SECTION(swift2_protocol_conformances)
+DECLARE_SWIFT_SECTION(swift2_type_metadata)
 
-// Add the section information in an image specified by an address in that
-// image.
-static void addBlockInImage(const InspectArgs *inspectArgs, const void *addr) {
-  const char *fname = nullptr;
-  if (addr) {
-    Dl_info info;
-    if (dladdr(addr, &info) == 0 || info.dli_fname == nullptr) {
-      return;
-    }
-    fname = info.dli_fname;
-  }
-  SectionInfo block = getSectionInfo(fname, inspectArgs->symbolName);
-  if (block.size > 0) {
-    inspectArgs->addBlock(block.data, block.size);
-  }
-}
-
-static void initializeSectionLookup(InspectArgs *inspectArgs) {
-  // Add section data in the main executable.
-  addBlockInImage(inspectArgs, nullptr);
-  // Search the loaded dls. This only searches the already
-  // loaded ones. Any images loaded after this are processed by
-  // addNewDSOImage() below.
-  dl_iterate_phdr(iteratePHDRCallback, reinterpret_cast<void *>(inspectArgs));
+#undef DECLARE_SWIFT_SECTION
 }
 
 void swift::initializeProtocolConformanceLookup() {
-  initializeSectionLookup(&ProtocolConformanceArgs);
+  const char *conformances = &__start_swift2_protocol_conformances;
+  if (size_t length = &__stop_swift2_protocol_conformances - conformances)
+    addImageProtocolConformanceBlockCallback(conformances, length);
 }
```

Comment:
Shouldn't the standard library conformances have already been registered by libswiftCore.so's own static constructor by this point?